In [3]:
# sempre a primeira célula do notebook, importar o que vc vai usar.
# se apareça erro aqui, a biblioteca não está instalada.
# solução: abra o terminal e rode: pip install pandas openpyxl


import pandas as pd                       # biblioteca de análise de dados
import re                                 # expressão regulares (para limpar texto)
from datetime import datetime             # para eu trabalhar com datas


# confirmar que as importações fucionaram

print(' pandas versão:', pd.__version__)
print(' Bibliotecas importadas com sucesso!')

 pandas versão: 2.3.3
 Bibliotecas importadas com sucesso!


In [4]:
# pd.read_csv()  # para ler arquivos .csv
# CAMINHO = onde está o arquivo. Como ele está na MESMA pasta que este notebook..................????????
# basta passar o nome do arquivo, entre aspas, para a função ler o arquivo e criar um DataFrame (tabela) com os dados.


CAMINHO = r"C:\Users\s2bcl\OneDrive\Documents\Lab365\aula_de_hoje\turma-visualizacao-de-dados\alunos\claudio_neves\semana_04\base_rh.csv"

df = pd.read_csv(
    CAMINHO,
    sep=';',                # separador de colunas é ';' e não ','
    encoding='cp1252',      # encoding do windows, para ler acentos
    decimal=','             # virgula como separado decimal
)

print('Dados carregados com sucesso!')
print(f' Total de linhas : {df.shape[0]}')
print(f' Total de colunas: {df.shape[1]}')

Dados carregados com sucesso!
 Total de linhas : 1000
 Total de colunas: 10


In [7]:
# df.head(n)  # para ver as primeiras linhas da tabela, e confirmar que os dados foram lidos corretamente.
# Padrão: df.head(5) mostra as 5 primeiras linhas, mas pode passar outro número para ver mais ou menos linhas. Se não passar nada, o padrão é 5.
# Use para ter uma rápida visão geral dos dados, e confirmar que as colunas foram lidas corretamente, e que os dados estão no formato esperado.

df.head(5)


,ID_Funcionario,Nome,Departamento,Cargo,Salario,Data_Admissao,Genero,Idade,Estado_Civil,Status
0,1,Julia Nunes,Logística,Coordenador,9088.34,13/08/2024,F,43,Solteiro,Inativo
1,2,Sr. Gustavo Duarte,TI,Gerente,8155.98,29/04/2017,F,59,Solteiro,Inativo
2,3,Srta. Mariana Cunha,RH,Coordenador,14027.93,11/12/2024,F,27,Divorciado,Ativo
3,4,Ana Sophia da Cruz,Produção,Técnico,6647.97,16/06/2019,M,50,Casado,Inativo
4,5,Dr. Nicolas Pinto,Produção,Gerente,8730.19,29/03/2019,F,62,Casado,Inativo


In [9]:
# df.tail(n)  # para ver as últimas linhas da tabela, e confirmar que os dados foram lidos corretamente.
# Serve para confirmar que o final da tabela também foi lido corretamente, e para ver se tem alguma informação importante no final da tabela (ex: rodapés, notas, etc).
# Se o arquivo tiver 1.000 linhas, o tail deve mostras os IDs (próximos de 1000) de 996 a 1.000, confirmando que todas as linhas foram lidas.


df.tail(10)

,ID_Funcionario,Nome,Departamento,Cargo,Salario,Data_Admissao,Genero,Idade,Estado_Civil,Status
990,991,Sr. Vicente Silveira,Produção,Gerente,8026.70,11/07/2021,F,55,Viúvo,Inativo
991,992,Emanuel da Cruz,Logística,Gerente,12572.11,30/01/2024,M,33,Viúvo,Ativo
992,993,Enrico Monteiro,RH,Coordenador,14677.33,26/12/2017,M,52,Divorciado,Inativo
993,994,Ana Lívia Cardoso,RH,Coordenador,14912.48,22/10/2015,M,27,Solteiro,Inativo
994,995,Fernando Monteiro,RH,Assistente,10983.02,09/01/2023,F,42,Viúvo,Ativo
995,996,Matheus Aragão,Financeiro,Analista,11389.03,06/05/2022,F,26,Viúvo,Ativo
996,997,Ryan Martins,Financeiro,Analista,14060.42,05/06/2023,F,37,Viúvo,Ativo
997,998,Isabelly Ferreira,TI,Gerente,2348.73,29/12/2023,F,58,Solteiro,Inativo
998,999,Thales Farias,Produção,Analista,9430.62,25/12/2019,M,41,Solteiro,Ativo
999,1000,Sra. Maysa Costela,Produção,Técnico,12549.72,14/11/2022,M,46,Viúvo,Inativo


In [10]:
# df.columns  # para ver os nomes das colunas, e confirmar que foram lidos corretamente, e que estão no formato esperado (ex: sem espaços, sem caracteres estranhos, etc).
# .tolist()  # para transformar o Index de colunas em uma lista, e facilitar a manipulação dos nomes das colunas (ex: para renomear, para acessar as colunas, etc).


print('Colunas do dataset: ')
print(df.columns.tolist())

Colunas do dataset: 
['ID_Funcionario', 'Nome', 'Departamento', 'Cargo', 'Salario', 'Data_Admissao', 'Genero', 'Idade', 'Estado_Civil', 'Status']


---
## Diagnóstico: tipos e nulos

**O que são tipos de dados?**  
Cada coluna armazena um tipo de informação. O Python precisa saber o tipo  
para saber o que pode fazer com ela:

| Tipo | O que é | Exemplo |
|---|---|---|
| `int64` | Número inteiro | 1, 42, 1000 |
| `float64` | Número decimal | 9088.34, 3.14 |
| `object` | Texto / string | "Julia Nunes", "Ativo" |
| `datetime64` | Data e hora | 2024-08-13 |

**O que são nulos?**  
São células vazias — o campo não foi preenchido no sistema de origem.  
Em Python/pandas, nulos aparecem como `NaN` (*Not a Number*) ou `NaT` (*Not a Time*).

In [15]:
# df.info() → resumo completo do DataFrame em uma só chamada:
# - Nome de cada coluna
# - Quantidade de valores não-nulos
# - Tipo de dado (dtype)
# - Uso de memória
# É o comando mais importante para o primeiro diagnóstico!


df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 10 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   ID_Funcionario  1000 non-null   int64  
 1   Nome            1000 non-null   object 
 2   Departamento    1000 non-null   object 
 3   Cargo           1000 non-null   object 
 4   Salario         1000 non-null   float64
 5   Data_Admissao   1000 non-null   object 
 6   Genero          1000 non-null   object 
 7   Idade           1000 non-null   int64  
 8   Estado_Civil    1000 non-null   object 
 9   Status          1000 non-null   object 
dtypes: float64(1), int64(2), object(7)
memory usage: 78.3+ KB


In [19]:
# df.isnull() → cria uma tabela de True/False (True = é nulo)
# .sum()      → soma os True de cada coluna (True vale 1, False vale 0)
# Resultado: quantidade de nulos por coluna


print('Quantidade de nulos por coluna: ')
print(df.isnull().sum())

print()



# percentual de nulo - mais útil em bases grandes, para entender a proporção de nulos em cada coluna.
# len(df) → quantidade total de linhas do DataFrame

print('Percentual de nulos por coluna: ')
print((df.isnull().sum() / len(df) * 100).round(2))


Quantidade de nulos por coluna: 
ID_Funcionario    0
Nome              0
Departamento      0
Cargo             0
Salario           0
Data_Admissao     0
Genero            0
Idade             0
Estado_Civil      0
Status            0
dtype: int64

Percentual de nulos por coluna: 
ID_Funcionario    0.0
Nome              0.0
Departamento      0.0
Cargo             0.0
Salario           0.0
Data_Admissao     0.0
Genero            0.0
Idade             0.0
Estado_Civil      0.0
Status            0.0
dtype: float64


---
## Estatísticas das colunas numéricas

`describe()` gera um resumo estatístico automático.  
Você vai usar isso para detectar valores suspeitos (ex: salário de R$0,00 ou idade de 200 anos).

| Estatística | O que significa |
|---|---|
| `count` | Quantos valores não-nulos existem |
| `mean` | Média |
| `std` | Desvio padrão — o quanto os valores variam em torno da média |
| `min` | Menor valor |
| `25%` | 25% dos registros estão abaixo desse valor |
| `50%` | Metade dos registros estão abaixo (mediana) |
| `75%` | 75% dos registros estão abaixo desse valor |
| `max` | Maior valor |

In [21]:
df.describe().round(2) # para ver estatísticas descritivas das colunas numéricas, e confirmar que os dados estão no formato esperado (ex: sem valores absurdos, sem erros de leitura, etc). 

,ID_Funcionario,Salario,Idade
count,1000.00,1000.00,1000.00
mean,500.50,8579.95,41.40
std,288.82,3657.37,13.67
min,1.00,2000.71,18.00
25%,250.75,5564.54,30.00
50%,500.50,8571.13,41.00
75%,750.25,11554.63,53.00
max,1000.00,14954.51,65.00


---
## 🗂️ CÉLULA 6 — Explorando colunas categóricas

**Coluna categórica** = coluna com poucos valores possíveis (departamento, cargo, status...)  

`nunique()` conta quantos valores diferentes existem.  
`value_counts()` mostra a distribuição — quantas vezes cada valor aparece.

In [22]:
# df.nunique() → retorna quantos valores ÚNICOS cada coluna tem
# Serve para identificar quais são categóricas (poucas opções)
# e quais são identificadoras (muitas opções diferentes)

print('Valores únicos por coluna:')
print(df.nunique())

Valores únicos por coluna:
ID_Funcionario    1000
Nome               976
Departamento         6
Cargo                5
Salario           1000
Data_Admissao      878
Genero               2
Idade               48
Estado_Civil         4
Status               2
dtype: int64


In [23]:
# df['coluna'].value_counts() → conta quantas vezes cada valor aparece,
# ordenado do mais frequente para o menos frequente

# Vamos ver todas as colunas categóricas de uma vez usando um loop:
# 'for col in lista' = para cada coluna na lista, executa o bloco abaixo

for col in ['Departamento', 'Cargo', 'Genero', 'Status', 'Estado_Civil']:
    print(f'\n── {col} ──')   # f-string: {} insere o valor da variável
    print(df[col].value_counts())


── Departamento ──
Departamento
Financeiro    189
Produção      182
RH            166
Vendas        160
Logística     156
TI            147
Name: count, dtype: int64

── Cargo ──
Cargo
Coordenador    225
Gerente        221
Assistente     199
Analista       178
Técnico        177
Name: count, dtype: int64

── Genero ──
Genero
F    510
M    490
Name: count, dtype: int64

── Status ──
Status
Ativo      515
Inativo    485
Name: count, dtype: int64

── Estado_Civil ──
Estado_Civil
Solteiro      265
Viúvo         261
Divorciado    240
Casado        234
Name: count, dtype: int64


---
## 🔎 CÉLULA 7 — Selecionando colunas e filtrando linhas

São as operações mais usadas no dia a dia. Você vai usar isso em todo projeto.

**Seleção de coluna** → `df['Nome']`  
**Seleção de múltiplas colunas** → `df[['Nome', 'Salario']]`  
**Filtro de linhas** → `df[df['Status'] == 'Ativo']`

In [28]:
# Selecionar UMA coluna → retorna uma Series (lista com índice)
# Uma Series é como uma coluna de planilha isolada

df['Departamento'].head(3)

0    Logística
1           TI
2           RH
Name: Departamento, dtype: object

In [29]:
# Selecionar MÚLTIPLAS colunas → retorna um DataFrame
# Note os dois colchetes: df[  ['col1', 'col2']  ]
# O primeiro [] é o operador de acesso do DataFrame
# O segundo [] é a lista de colunas

df[['Nome', 'Departamento', 'Cargo', 'Salario']].head(6)

,Nome,Departamento,Cargo,Salario
0,Julia Nunes,Logística,Coordenador,9088.34
1,Sr. Gustavo Duarte,TI,Gerente,8155.98
2,Srta. Mariana Cunha,RH,Coordenador,14027.93
3,Ana Sophia da Cruz,Produção,Técnico,6647.97
4,Dr. Nicolas Pinto,Produção,Gerente,8730.19
5,Marcela Costa,RH,Analista,12403.42


In [30]:
# FILTRO SIMPLES: selecionar só as linhas onde Status == 'Ativo'
# df['Status'] == 'Ativo' → cria uma Series de True/False
# df[ True/False ]        → mantém só as linhas onde é True

ativos = df[df['Status'] == 'Ativo']

print(f'Total de funcionários : {len(df)}')
print(f'Funcionários ativos   : {len(ativos)}')
print(f'Funcionários inativos : {len(df) - len(ativos)}')

Total de funcionários : 1000
Funcionários ativos   : 515
Funcionários inativos : 485


In [32]:
# FILTRO COMPOSTO: duas condições ao mesmo tempo
# Use & para AND (as duas condições precisam ser verdadeiras)
# Use | para OR  (pelo menos uma precisa ser verdadeira)
# IMPORTANTE: cada condição precisa estar entre parênteses

gerentes_ti = df[
    (df['Cargo'] == 'Gerente') & (df['Departamento'] == 'TI')
]

print(f'Gerentes na área de TI: {len(gerentes_ti)}')
print()
gerentes_ti[['Nome', 'Salario', 'Status', 'Data_Admissao']]

Gerentes na área de TI: 31



,Nome,Salario,Status,Data_Admissao
1,Sr. Gustavo Duarte,8155.98,Inativo,29/04/2017
49,Benjamin da Costa,9766.01,Inativo,27/04/2024
130,Catarina Costela,9597.83,Inativo,03/07/2023
131,Laura Cardoso,7364.46,Inativo,29/06/2018
175,Maria Cardoso,12127.27,Inativo,22/05/2024
187,Ana Júlia Cardoso,8738.63,Ativo,30/05/2021
247,Ian Pereira,2395.32,Ativo,10/05/2019
273,Ana Clara Peixoto,7301.14,Ativo,16/11/2019
296,Augusto Santos,8237.80,Inativo,26/08/2021
307,Manuela Barbosa,7015.91,Ativo,21/06/2017


In [33]:
# Estatísticas rápidas de uma coluna numérica
# .mean()  → média
# .max()   → maior valor
# .min()   → menor valor
# .median()→ mediana (valor do meio quando ordenados)

print(f"Salário médio : R$ {df['Salario'].mean():,.2f}")
print(f"Salário máximo: R$ {df['Salario'].max():,.2f}")
print(f"Salário mínimo: R$ {df['Salario'].min():,.2f}")
print(f"Salário mediana:R$ {df['Salario'].median():,.2f}")

# :,.2f → formata o número com separador de milhar (,) e 2 casas decimais (.2f)

Salário médio : R$ 8,579.95
Salário máximo: R$ 14,954.51
Salário mínimo: R$ 2,000.71
Salário mediana:R$ 8,571.13
